js/tools/tooloption.js

```diff
@@ -0,0 +1,113 @@
+/**
+ * Class for generating options for tools. Types of options are specified as string arguments in the constructor
+ * and mapped to functions that return DOM elements that are displayed in the tool customization panel on the right.
+ * This object is intended to be applied in an "options" key on the class representing that tool.
+ * 
+ * For example:
+ * rectangle.js contains the Rectangle class. At the very bottom, Rectangle.options is set to a new ToolOption
+ * with "fill" and "stroke" passed into its constructor. Any associated values in the Rectangle class that needs
+ * to be altered with those options can be set to values in the Rectangle.options object.
+ */
+
+class ToolOption {
+    constructor(...options) {
+        this.options = options;
+        this.optionMapping = {
+            "fill": () => this.fillOptions(),
+            "stroke": () => this.strokeOptions()
+        }
+    }
+
+    fillOptions() {
+        this.fillSettings = {}; // Settings about the fill of a tool are stored in {ToolClass}.objects.fillSettings
+        this.fillSettings["color"] = "#000000";
+        var colorList = document.createElement("div");
+        colorList.style.display = "flex";
+        colorList.style.flexWrap = "wrap";
+
+        var colors = [
```

Comment:
would be more flexible and easier to insert themes if we have an input element ad you can add onfocus and onblur listeners. Maybe should a preview/confirmed cell.

Reply:
also have a transparent option i believe most fills should be transparent, this comes for free with input

Reply:
> would be more flexible and easier to insert themes if we have an input element ad you can add onfocus and onblur listeners. Maybe should a preview/confirmed cell.

So a text box that allows users to add custom colors?